### PR TITLE
(CDAP-13711) Runtime Monitor State Management

### PR DESCRIPTION
**Pull request by chtyim**
Please review by commits in the commit order.

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java [hunk 1]
```diff
@@ -238,23 +245,24 @@ public TwillPreparer prepare(TwillApplication application) {
     SSHKeyPair sshKeyPair = createSSHKeyPair(keysDir, cluster);
 
     // Creates and return the twill preparer
-    String remoteHost = Preconditions.checkNotNull(masterNode.getProperties().get("ip.external"));
+    SSHConfig sshConfig = SSHConfig.builder(masterNode.getProperties().get("ip.external"))
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: we moved this string to a Constants.Node class in the runtime spi, should use that.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteRuntimeDataset.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+   * @param lastProgramRunId an optional {@link ProgramRunId} returned from previous scan to continue the scanning
+   * @return a {@link List} of {@link ProgramRunId} to the corresponding {@link ProgramOptions} for each run
+   */
+  public List<Map.Entry<ProgramRunId, ProgramOptions>> scan(int limit, @Nullable ProgramRunId lastProgramRunId) {
+    MDSKey stopKey = new MDSKey(Bytes.stopKeyForPrefix(new MDSKey.Builder().add(KEY_PREFIX).build().getKey()));
+    MDSKey startKey;
+
+    if (lastProgramRunId == null) {
+      // If there is no start program run id, just scan from the state prefix
+      startKey = new MDSKey.Builder().add(KEY_PREFIX).build();
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteRuntimeDataset.java]: this is already done to generate the stop key, better to just do it the one time and only check if lastProgramRunId is set.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java [hunk 3]
```diff
@@ -356,4 +400,161 @@ private void saveKeyStores(KeyStore serverKeyStore, KeyStore clientKeyStore, Loc
...
+      while (!completed) {
+        // Scan the dataset and update the controllers map. Retry on all exception.
+        completed = Retries.callWithRetries(() -> Transactionals.execute(transactional, context -> {
+          RemoteRuntimeDataset dataset = RemoteRuntimeDataset.create(context, datasetFramework);
+          List<Map.Entry<ProgramRunId, ProgramOptions>> scanResult = dataset.scan(limit, lastProgramRunId.get());
+          for (Map.Entry<ProgramRunId, ProgramOptions> entry : scanResult) {
+            ProgramRunId programRunId = entry.getKey();
+            ProgramOptions programOptions = entry.getValue();
+
+            if (RunIds.getTime(programRunId.getRun(), TimeUnit.MILLISECONDS) > startMillis) {
```
Comment (albertshau) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: Seems like we should set lastProgramRunId earlier. If we scan a batch containing only these types of rows, we'll infinitely loop.

Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: Right. good catch.

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteRuntimeDataset.java (hunk 2 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteRuntimeDataset.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java (hunk 1 above)
Comment (chtyim) [reply cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java]: fixed

**Comment by chtyim** [comment]
Build https://builds.cask.co/browse/CDAP-RUT1579

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by chtyim** [comment]
Squashed commits and waiting for build

**Comment by chtyim** [comment]
Build passed, merging

**Pull request merged**